server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -1126,4 +1137,50 @@ public void modeChanged(Mode mode) {
             }
         }
     }
+
+    /**
+     * Lookup job status by job id.
+     *
+     * @param jobId
+     *  Id of the requested job
+     *
+     * @return
+     *  {@link AsyncJobStatus} of the requested job
+     *
+     * @throws JobAccessException.NotFound
+     *  When requested job was not found
+     */
+    public AsyncJobStatus get(final String jobId) throws JobAccessException {
+        final AsyncJobStatus status = this.jobCurator.get(jobId);
+        if (status == null) {
+            throw new JobAccessException.NotFound(jobId);
+        }
+        return status;
+    }
+
+    /**
+     * Request cancellation of the job with the given id.
+     *
+     * @param jobId
+     *  Id of the cancelled job
+     *
+     * @return
+     *  {@link AsyncJobStatus} of the cancelled job
+     *
+     * @throws JobAccessException.NotFound
+     *  When requested job was not found
+     *
+     * @throws JobAccessException.InvalidState
+     *  When cancelled job is already terminated (CANCELLED, COMPLETED, ...)
+     */
+    public AsyncJobStatus cancel(final String jobId) throws JobAccessException {
+        final AsyncJobStatus status = this.get(jobId);
+        if (status.getState().isTerminal()) {
+            throw new JobAccessException.InvalidState(jobId, status.getState());
+        }
+        status.setState(JobState.CANCELED);
```

Comment:
This will work for this PR, but there's more work to be done here. There are a number of cases here which won't work as the operation implies, and those may or may not require attention. At the very least, they should be documented in the javadocs so we minimize confusion in the future.

Reply:
What cases do you mean?

Reply:
Off the top of my head:
- Job is currently executing, potentially on another node
- Created -> cancelled during message sending or failure will lead to created -> failed at the same time
- General race condition where another node (or even the current node) has started a transaction which will process the job message while this node is starting to cancel it (queued, failed_with_retry -> executing)